### PR TITLE
Spiral/aneksteind

### PR DESCRIPTION
**Pull request by oinoom**

### CODE REVIEW

File: week4/spiral/spiral.hs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+module Spiral where
+
+import Data.Matrix
+
+data Direction = R | D | U | L
+
+main :: IO ()
+main = do
+    putStrLn("Please enter a size for the spiral!")
+    i <- getInt
```
Comment (balachandranc) [inline week4/spiral/spiral.hs:10]: You can avoid writing "getInt" function. You could do:
`i <- read <$> getLine`
The compiler will infer the type of `read` by how the value is used down the line.

**Review by balachandranc — COMMENTED** [review]
Minor comments.

File: week4/spiral/spiral.hs (hunk 1 above)
Comment (oinoom) [reply week4/spiral/spiral.hs:10]: ah interesting @balachandranc, thank you

**Comment by 444c43** [comment]
using Data.Matrix is a pretty interesting approach. i didn't even know this library existed.

odd number sizes seem to look better than even as the inside ends up as a grouping of 2x2 stars. for instance, spiral 4 looks like:

```
* * * * 
      * 
* *   * 
* * * * 
```

by adding an additional row to your matrix.

```
* * * * 
      * 
* *   * 
*     * 
* * * * 
```

i'm not sure the `Direction = R | D | U | L` approach would work so cleanly though.

regardless, i tried it out and it works!

**Review by 444c43 — APPROVED** [review]

File: week4/spiral/spiral.hs (hunk 1 above)
Comment (int-index) [reply week4/spiral/spiral.hs:10]: `read <$> getLine` is called `readLn`:

```
Prelude> :t readLn 
readLn :: Read a => IO a
```

File: week4/spiral/spiral.hs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+module Spiral where
+
+import Data.Matrix
+
+data Direction = R | D | U | L
+
+main :: IO ()
+main = do
+    putStrLn("Please enter a size for the spiral!")
```
Comment (neongreen) [inline week4/spiral/spiral.hs:9]: Unnecessary parens

File: week4/spiral/spiral.hs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    i <- getInt
+    putStrLn $ concatMap (\line -> (concatMap (\x -> if x == 1 then "* " else "  ") line) ++ "\n") $ toLists $ spiral i 
+
+spiral :: Int -> Matrix Int
+spiral i = buildSpiral (zero i i) i R (1,1) 0
+
+buildSpiral :: Matrix Int -> Int -> Direction -> (Int, Int) -> Int -> Matrix Int
+buildSpiral m i dir curr count | count >=3 = (setElem 1 curr m)
```
Comment (neongreen) [inline week4/spiral/spiral.hs:17]: Unnecessary parens

File: week4/spiral/spiral.hs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+                               | (available m curr) =
```
Comment (neongreen) [inline week4/spiral/spiral.hs:18]: Unnecessary parens

File: week4/spiral/spiral.hs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    if (atBound dir curr i m)
```
Comment (neongreen) [inline week4/spiral/spiral.hs:19]: Unnecessary parens

File: week4/spiral/spiral.hs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    then buildSpiral m i (nextDir dir) curr (count + 1)
+    else buildSpiral (setElem 1 curr m) i dir (nextPos dir curr) 0
+                         | otherwise = m
+
+available :: Matrix Int -> (Int, Int) -> Bool
+available m pos = if (m ! pos) == 1 then False else True
```
Comment (neongreen) [inline week4/spiral/spiral.hs:25]: `available m pos = m ! pos /= 1`

**Review by neongreen — CHANGES_REQUESTED** [review]
It could've been done quite a bit easier – see @vitcra's and @balachandranc's solutions. Also, put the spiral.hs file into a directory called `aneksteind` before merging.

**Comment by neongreen** [comment]
(To be clear, only “put the spiral.hs file into a directory called aneksteind before merging” is a requested change.)

**Pull request merged**